'use client'
import { jsx } from 'react/jsx-runtime';
import { createContext, useContext, useState, useCallback, useMemo } from 'react';
import { useInternalDynamicEvents } from '../../utils/hooks/events/useDynamicEvents/useDynamicEvents.js';
import { useErrorContext } from '../ErrorContext/ErrorContext.js';

const ViewContext = createContext(undefined);
const ViewContextProvider = ({ children, initialView = { props: { isWalletConnectList: true }, type: 'wallet-list' }, }) => {
    const [_view, _setView] = useState(null);
    const view = _view || initialView;
    const { clearError } = useErrorContext();
    const goToInitialView = useCallback(() => {
        _setView(null);
    }, []);
    const setView = useCallback((type, props, events) => {
        // we need to clear errors before proceeding to the next view
        // but only if the next view is different from the current view
        // this is to prevent the error from being cleared when the same view is re-rendered
        if (view.type !== type)
            clearError();
        _setView({ events, props, type });
    }, []);
    useInternalDynamicEvents('logout', () => goToInitialView());
    const value = useMemo(() => ({
        goToInitialView,
        setView,
        view,
    }), [goToInitialView, setView, view]);
    return jsx(ViewContext.Provider, { value: value, children: children });
};
const useViewContext = () => {
    const context = useContext(ViewContext);
    if (context === undefined) {
        throw new Error('usage of useViewContext not wrapped in `ViewContextProvider`.');
    }
    return context;
};

export { ViewContext, ViewContextProvider, useViewContext };
